src/extra/stream.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2015 The Gfx-rs Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Render stream extension. Stream is something you can send batches to.
+//! It includes a renderer and an output, stored by constrained types.
+
+
+use device::{Device, InstanceCount, Resources, VertexCount};
+use device::draw::CommandBuffer;
+use device::target::{ClearData, Mask, Mirror, Rect};
+use render::{DrawError, Renderer, RenderFactory};
+use render::batch::Batch;
+use render::target::Output;
+
+//TODO: when `Stream` gets adopted, we can remove these methods:
+//  - Renderer::draw(...)
+//  - Renderer::draw_instanced(...)
+// in favor of bare `Renderer::draw_all` (renamed), since most
+// of the usage will go through `Stream`, which already has this.
+
+/// Render stream abstraction.
+pub trait Stream<R: Resources> {
+    /// Command buffer type to constraint the `Renderer`.
+    type CommandBuffer: CommandBuffer<R>;
+    /// Constrained `Output` type.
+    type Output: Output<R>;
+
+    /// Get the output only.
+    fn get_output(&self) -> &Self::Output;
+
+    /// Access both of the stream components.
+    fn access(&mut self) -> (&mut Renderer<R, Self::CommandBuffer>, &Self::Output);
```

Comment:
`get_access_mut` might be better here, but I'm not sure. Discussion on get/set patterns for Piston is here https://github.com/PistonDevelopers/piston/issues/882